utreexo/rev.go

```diff
@@ -0,0 +1,352 @@
+package utreexo
+
+import (
+	"encoding/binary"
+	"errors"
+	"io"
+	"math"
+	"math/big"
+	"os"
+)
+
+// RevReader ...
+type RevReader struct {
+	Records []*BlockRecord
+}
+
+// BlockRecord ...
+type BlockRecord struct {
+	Magic uint32
+	Size  uint32
+	Data  []byte
+	Block *CBlockUndo
+	Hash  []byte
+}
+
+// CBlockUndo ...
+type CBlockUndo struct {
+	Txs []*CTxUndo
+}
+
+// CTxUndo ...
+type CTxUndo struct {
+	Ins []*CTxInUndo
+}
+
+// CTxInUndo ...
+type CTxInUndo struct {
+	Height uint64
+	Script []byte
+	Amount uint64
+}
+
+// OpenRevReader ...
+func OpenRevReader(name string) (*RevReader, error) {
+	f, err := os.Open(name)
+	if err != nil {
+		return nil, err
+	}
+	fi, err := f.Stat()
+	if err != nil {
+		f.Close()
+		return nil, err
+	}
+	r := new(RevReader)
+	err = r.init(f, fi.Size())
+	if err != nil {
+		f.Close()
+		return nil, err
+	}
+	return r, nil
+}
+
+// NewRevReader ...
+func NewRevReader(r io.ReaderAt, size int64) (*RevReader, error) {
+	if size < 0 {
+		return nil, errors.New("rev: size cannot be negative")
+	}
+	rr := new(RevReader)
+	err := rr.init(r, size)
+	if err != nil {
+		return nil, err
+	}
+	return rr, nil
+}
+
+func (rr *RevReader) init(r io.ReaderAt, size int64) error {
+	pos := int64(0)
+	for size > pos {
+		record := new(BlockRecord)
+		buf := make([]byte, 4)
+		_, err := r.ReadAt(buf, pos)
+		if err != nil {
+			return err
+		}
+		record.Magic = binary.LittleEndian.Uint32(buf)
+		if record.Magic == 0 {
```

Comment:
Here, since we know the magic bytes of testnet and mainnet, you could hold a test to see if it matches.

Reply:
The current network is limited, but there are no restrictions for Magic (mainnet, testnet, regtestnet, signet, etc ...), so I decided to remove this check.